zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/Leader.java

```diff
@@ -271,39 +278,42 @@ public boolean isQuorumSynced(QuorumVerifier qv) {
        return qv.containsQuorum(ids);
     }
 
-    private final ServerSocket ss;
+    private final List<ServerSocket> serverSockets = new LinkedList<>();
 
     Leader(QuorumPeer self,LeaderZooKeeperServer zk) throws IOException {
         this.self = self;
         this.proposalStats = new BufferStats();
+
+        Set<InetSocketAddress> addresses;
+        if (self.getQuorumListenOnAllIPs()) {
+            addresses = self.getQuorumAddress().getWildcardAddresses();
+        } else {
+            addresses = self.getQuorumAddress().getAllAddresses();
+        }
+
+        for (InetSocketAddress address : addresses) {
+            serverSockets.add(createServerSocket(address, self.shouldUsePortUnification(), self.isSslQuorum()));
+        }
+
+        this.zk = zk;
+    }
+
+    ServerSocket createServerSocket(InetSocketAddress address, boolean portUnification, boolean sslQuorum)
+            throws IOException {
+        ServerSocket serverSocket;
         try {
-            if (self.shouldUsePortUnification() || self.isSslQuorum()) {
-                boolean allowInsecureConnection = self.shouldUsePortUnification();
-                if (self.getQuorumListenOnAllIPs()) {
-                    ss = new UnifiedServerSocket(self.getX509Util(), allowInsecureConnection, self.getQuorumAddress().getPort());
-                } else {
-                    ss = new UnifiedServerSocket(self.getX509Util(), allowInsecureConnection);
-                }
+            if (portUnification || sslQuorum) {
+                serverSocket = new UnifiedServerSocket(self.getX509Util(), portUnification);
             } else {
-                if (self.getQuorumListenOnAllIPs()) {
-                    ss = new ServerSocket(self.getQuorumAddress().getPort());
-                } else {
-                    ss = new ServerSocket();
-                }
-            }
-            ss.setReuseAddress(true);
-            if (!self.getQuorumListenOnAllIPs()) {
-                ss.bind(self.getQuorumAddress());
+                serverSocket = new ServerSocket();
             }
+            serverSocket.setReuseAddress(true);
+            serverSocket.bind(address);
+            return serverSocket;
         } catch (BindException e) {
-            if (self.getQuorumListenOnAllIPs()) {
-                LOG.error("Couldn't bind to port " + self.getQuorumAddress().getPort(), e);
-            } else {
-                LOG.error("Couldn't bind to " + self.getQuorumAddress(), e);
-            }
+            LOG.error("Couldn't bind to " + self.getQuorumAddress(), e);
```

Comment:
Instead of `self.getQuorumAddress()`, you should print `address`, because that's the actual address that you're trying to bind.

Reply:
thanks!